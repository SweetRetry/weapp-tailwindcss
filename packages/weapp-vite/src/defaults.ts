import { addExtension, removeExtension } from '@weapp-core/shared'
import { readPackageJSON } from 'pkg-types'
import path from 'pathe'
import tsconfigPaths from 'vite-tsconfig-paths'
import type { UserConfig, WatchOptions } from './types'
import { vitePluginWeapp } from './plugins'
import type { Context } from './context'

export const defaultExcluded: string[] = ['**/node_modules/**', '**/miniprogram_npm/**']

export function getWeappWatchOptions(): WatchOptions {
  return {
    paths: ['**/*.{wxml,json,wxs}', '**/*.{png,jpg,jpeg,gif,svg,webp}', '.env', '.env.*'],
    ignored: [
      ...defaultExcluded,
    ],
  }
}

export async function getDefaultViteConfig(ctx: Context): Promise<UserConfig> {
  const localPackageJson = await readPackageJSON()
  const external: string[] = []
  if (localPackageJson.dependencies) {
    external.push(...Object.keys(localPackageJson.dependencies))
  }

  return {
    mode: ctx.options.mode,
    build: {
      rollupOptions: {
        output: {
          format: 'cjs',
          strict: false,
          entryFileNames: (chunkInfo) => {
            const name = ctx.relativeSrcRoot(chunkInfo.name)
            if (name.endsWith('.ts')) {
              const baseFileName = removeExtension(name)
              if (baseFileName.endsWith('.wxs')) {
                return path.normalize((baseFileName))
              }
              return path.normalize(addExtension(baseFileName, '.js'))
            }
            return path.normalize(name)
          },
        },
        external,
      },
      assetsDir: '.',
      commonjsOptions: {
        transformMixedEsModules: true,
        include: undefined,
      },
    },
    plugins: [
      vitePluginWeapp(ctx),
      tsconfigPaths(),
    ],
  }
}
