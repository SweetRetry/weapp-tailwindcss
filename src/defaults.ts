import defu from 'defu'
import type { UserDefinedOptions } from './types'
// import { mangleClassRegex } from '@/mangle/expose'
const noop = () => {}

export const defaultOptions: Required<UserDefinedOptions> = {
  cssMatcher: (file) => /.+\.(?:wx|ac|jx|tt|q|c)ss$/.test(file),
  htmlMatcher: (file) => /.+\.(?:(?:(?:wx|ax|jx|ks|tt|q)ml)|swan)$/.test(file),
  jsMatcher: (file) => {
    if (file.includes('node_modules')) {
      return false
    }
    return /.+\.[jt]sx?$/.test(file)
  },
  mainCssChunkMatcher: (file, appType) => {
    switch (appType) {
      case 'uni-app': {
        return /^common\/main/.test(file)
      }
      case 'mpx': {
        return /^app/.test(file)
      }
      case 'taro': {
        return /^app/.test(file)
      }
      case 'remax': {
        return /^app/.test(file)
      }
      case 'rax': {
        return /^bundle/.test(file)
      }
      case 'native': {
        return /^app/.test(file)
      }
      case 'kbone': {
        return /^(?:common\/)?miniprogram-app/.test(file)
      }
      default: {
        return true
      }
    }
  },
  cssPreflight: {
    'box-sizing': 'border-box',
    'border-width': '0',
    'border-style': 'solid',
    'border-color': 'currentColor'
  },
  cssPreflightRange: 'view',
  replaceUniversalSelectorWith: 'view',
  disabled: false,
  customRuleCallback: noop,
  onLoad: noop,
  onStart: noop,
  onEnd: noop,
  onUpdate: noop,
  mangle: false,
  framework: 'react'

  // onBeforeUpdate: noop
}

export function getOptions (options: UserDefinedOptions = {}): Required<UserDefinedOptions> {
  if (options.mangle === true) {
    // https://uniapp.dcloud.net.cn/tutorial/miniprogram-subject.html#%E5%B0%8F%E7%A8%8B%E5%BA%8F%E8%87%AA%E5%AE%9A%E4%B9%89%E7%BB%84%E4%BB%B6%E6%94%AF%E6%8C%81
    options.mangle = {
      exclude: [/node[-_]modules/, /(wx|my|swan|tt|ks|jd)components/]
    }
  } else if (typeof options.mangle === 'object') {
    if (!Array.isArray(options.mangle)) {
      options.mangle.exclude = [/node[-_]modules/, /(wx|my|swan|tt|ks|jd)components/]
    }
  }
  return defu(options, defaultOptions)
}
