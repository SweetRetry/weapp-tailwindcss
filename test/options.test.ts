import { getOptions } from '@/defaults'
describe('get options', () => {
  it('default options', () => {
    const options = getOptions()
    expect(options).toMatchSnapshot()
  })
  it('vue framework', () => {
    const options = getOptions({
      framework: 'vue'
    })
    expect(options.framework).toBe('vue2')
    expect(options).toMatchSnapshot()
  })

  it('default matcher', () => {
    const { cssMatcher, jsMatcher, mainCssChunkMatcher, htmlMatcher } = getOptions()
    expect(cssMatcher('a.css')).toBe(true)
    expect(jsMatcher('a.js')).toBe(true)
    expect(jsMatcher('node_modules/a.js')).toBe(false)
    expect(mainCssChunkMatcher('app.wxss', 'native')).toBe(true)
    expect(htmlMatcher('a.wxml')).toBe(true)
  })

  it('glob matcher', () => {
    const { cssMatcher, jsMatcher, mainCssChunkMatcher, htmlMatcher } = getOptions({
      cssMatcher: '*.xxss',
      jsMatcher: '*.abcd',
      mainCssChunkMatcher: '*.main',
      htmlMatcher: ['*.wxmm', '*.plmm']
    })
    expect(cssMatcher('a.xxss')).toBe(true)
    expect(jsMatcher('a.abcd')).toBe(true)
    expect(mainCssChunkMatcher('app.main', 'native')).toBe(true)
    expect(htmlMatcher('a.wxmm')).toBe(true)
    expect(htmlMatcher('a.plmm')).toBe(true)
  })
})
