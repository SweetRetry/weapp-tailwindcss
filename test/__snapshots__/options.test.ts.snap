// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`get options default options 1`] = `
{
  "cssMatcher": [Function],
  "cssPreflight": {
    "border-color": "currentColor",
    "border-style": "solid",
    "border-width": "0",
    "box-sizing": "border-box",
  },
  "cssPreflightRange": "view",
  "customAttributes": {},
  "customReplaceDictionary": {
    "!": "_i_",
    """: "_dq_",
    "#": "_h_",
    "$": "_do_",
    "%": "_p_",
    "&": "_am_",
    "'": "_q_",
    "(": "_pl_",
    ")": "_qr_",
    "*": "_a_",
    "+": "_plus_",
    ",": "_co_",
    ".": "_d_",
    "/": "_s_",
    ":": "_c_",
    ";": "_se_",
    "<": "_l_",
    "=": "_e_",
    ">": "_g_",
    "?": "_qu_",
    "@": "_at_",
    "[": "_bl_",
    "\\": "_bs_",
    "]": "_br_",
    "^": "_ca_",
    "\`": "_bq_",
    "{": "_bal_",
    "|": "_b_",
    "}": "_bar_",
    "~": "_t_",
  },
  "customRuleCallback": [Function],
  "disabled": false,
  "escapeEntries": [
    [
      "[",
      "_bl_",
    ],
    [
      "]",
      "_br_",
    ],
    [
      "(",
      "_pl_",
    ],
    [
      ")",
      "_qr_",
    ],
    [
      "#",
      "_h_",
    ],
    [
      "!",
      "_i_",
    ],
    [
      "/",
      "_s_",
    ],
    [
      "\\",
      "_bs_",
    ],
    [
      ".",
      "_d_",
    ],
    [
      ":",
      "_c_",
    ],
    [
      "%",
      "_p_",
    ],
    [
      ",",
      "_co_",
    ],
    [
      "'",
      "_q_",
    ],
    [
      """,
      "_dq_",
    ],
    [
      "*",
      "_a_",
    ],
    [
      "&",
      "_am_",
    ],
    [
      "@",
      "_at_",
    ],
    [
      "{",
      "_bal_",
    ],
    [
      "}",
      "_bar_",
    ],
    [
      "+",
      "_plus_",
    ],
    [
      ";",
      "_se_",
    ],
    [
      "<",
      "_l_",
    ],
    [
      "~",
      "_t_",
    ],
    [
      "=",
      "_e_",
    ],
    [
      ">",
      "_g_",
    ],
    [
      "?",
      "_qu_",
    ],
    [
      "^",
      "_ca_",
    ],
    [
      "\`",
      "_bq_",
    ],
    [
      "|",
      "_b_",
    ],
    [
      "$",
      "_do_",
    ],
  ],
  "framework": "react",
  "htmlMatcher": [Function],
  "jsMatcher": [Function],
  "jsxHandler": [Function],
  "jsxRenameLoaderPath": "",
  "loaderOptions": {
    "jsxRename": false,
  },
  "mainCssChunkMatcher": [Function],
  "mangle": false,
  "onEnd": [Function],
  "onLoad": [Function],
  "onStart": [Function],
  "onUpdate": [Function],
  "replaceUniversalSelectorWith": "view",
  "styleHandler": [Function],
  "templeteHandler": [Function],
}
`;

exports[`get options vue framework 1`] = `
{
  "cssMatcher": [Function],
  "cssPreflight": {
    "border-color": "currentColor",
    "border-style": "solid",
    "border-width": "0",
    "box-sizing": "border-box",
  },
  "cssPreflightRange": "view",
  "customAttributes": {},
  "customReplaceDictionary": {
    "!": "_i_",
    """: "_dq_",
    "#": "_h_",
    "$": "_do_",
    "%": "_p_",
    "&": "_am_",
    "'": "_q_",
    "(": "_pl_",
    ")": "_qr_",
    "*": "_a_",
    "+": "_plus_",
    ",": "_co_",
    ".": "_d_",
    "/": "_s_",
    ":": "_c_",
    ";": "_se_",
    "<": "_l_",
    "=": "_e_",
    ">": "_g_",
    "?": "_qu_",
    "@": "_at_",
    "[": "_bl_",
    "\\": "_bs_",
    "]": "_br_",
    "^": "_ca_",
    "\`": "_bq_",
    "{": "_bal_",
    "|": "_b_",
    "}": "_bar_",
    "~": "_t_",
  },
  "customRuleCallback": [Function],
  "disabled": false,
  "escapeEntries": [
    [
      "[",
      "_bl_",
    ],
    [
      "]",
      "_br_",
    ],
    [
      "(",
      "_pl_",
    ],
    [
      ")",
      "_qr_",
    ],
    [
      "#",
      "_h_",
    ],
    [
      "!",
      "_i_",
    ],
    [
      "/",
      "_s_",
    ],
    [
      "\\",
      "_bs_",
    ],
    [
      ".",
      "_d_",
    ],
    [
      ":",
      "_c_",
    ],
    [
      "%",
      "_p_",
    ],
    [
      ",",
      "_co_",
    ],
    [
      "'",
      "_q_",
    ],
    [
      """,
      "_dq_",
    ],
    [
      "*",
      "_a_",
    ],
    [
      "&",
      "_am_",
    ],
    [
      "@",
      "_at_",
    ],
    [
      "{",
      "_bal_",
    ],
    [
      "}",
      "_bar_",
    ],
    [
      "+",
      "_plus_",
    ],
    [
      ";",
      "_se_",
    ],
    [
      "<",
      "_l_",
    ],
    [
      "~",
      "_t_",
    ],
    [
      "=",
      "_e_",
    ],
    [
      ">",
      "_g_",
    ],
    [
      "?",
      "_qu_",
    ],
    [
      "^",
      "_ca_",
    ],
    [
      "\`",
      "_bq_",
    ],
    [
      "|",
      "_b_",
    ],
    [
      "$",
      "_do_",
    ],
  ],
  "framework": "vue2",
  "htmlMatcher": [Function],
  "jsMatcher": [Function],
  "jsxHandler": [Function],
  "jsxRenameLoaderPath": "",
  "loaderOptions": {
    "jsxRename": false,
  },
  "mainCssChunkMatcher": [Function],
  "mangle": false,
  "onEnd": [Function],
  "onLoad": [Function],
  "onStart": [Function],
  "onUpdate": [Function],
  "replaceUniversalSelectorWith": "view",
  "styleHandler": [Function],
  "templeteHandler": [Function],
}
`;
